src/main/kotlin/org/rust/ide/hints/RsGenericParameterInfoHandler.kt

```diff
@@ -0,0 +1,140 @@
+        context.highlightedElement = null
+        context.showHint(element, element.textRange.startOffset, this)
+    }
+
+    // TODO: don't disappear in nested generic types
+    override fun updateParameterInfo(parameterOwner: RsTypeArgumentList, context: UpdateParameterInfoContext) {
+        curParam = getCurrentParameterIndex(parameterOwner.node, context.offset, RsElementTypes.COMMA)
+        when {
+            context.parameterOwner == null -> context.parameterOwner = parameterOwner
+
+            // occurs e.g. in case of up-down cursor moving
+            context.parameterOwner != parameterOwner -> context.removeHint()
+        }
+    }
+
+    override fun updateUI(p: HintLine, context: ParameterInfoUIContext) {
+        hintText = p.presentText
+        context.setupUIComponentPresentation(
+            hintText,
+            p.getRange(curParam).startOffset,
+            p.getRange(curParam).endOffset,
+            false, // define whole hint line grayed
+            false,
+            false, // define grayed part of args before highlight
+            context.defaultParameterColor
+        )
+    }
+
+    override fun getParametersForLookup(item: LookupElement?, context: ParameterInfoContext?): Array<Any>? = null
+
+    override fun couldShowInLookup() = false
+
+    override fun findElementForUpdatingParameterInfo(context: UpdateParameterInfoContext): RsTypeArgumentList? =
+        findExceptColonColon(context)
+
+    override fun findElementForParameterInfo(context: CreateParameterInfoContext): RsTypeArgumentList? {
+        val parameterList = findExceptColonColon(context) ?: return null
+        val parent = parameterList.parent
+        val genericDeclaration = if (parent is RsMethodCall || parent is RsPath) {
+            parent.reference?.resolve()  as? RsGenericDeclaration ?: return null
+        } else {
+            return null
+        }
+        val typesWithBounds = genericDeclaration.typeParameters.nullize() ?: return null
+        val presenter = RsGenericPresenter(typesWithBounds)
+        context.itemsToShow = listOfNotNull(presenter.firstLine, presenter.secondLine).toTypedArray()
+        return parameterList
+    }
+
+    // to avoid hint on :: before <>
+    private fun findExceptColonColon(context: ParameterInfoContext?): RsTypeArgumentList? {
+        val element = context?.file?.findElementAt(context.editor.caretModel.offset) ?: return null
+        if (element.elementType == RsElementTypes.COLONCOLON) return null
+        return element.ancestorStrict()
+    }
+}
+
+/**
+ * Stores the text representation and ranges for parameters
+ */
+class HintLine(
+    val presentText: String,
+    val ranges: List<TextRange>
+) {
+    fun getRange(index: Int) = if (index !in 0 until ranges.size) TextRange.EMPTY_RANGE else ranges[index]
+}
+
+/**
+ * Calculates the text representation and ranges for parameters
+ */
+class RsGenericPresenter(
```

Comment:
This class isn't needed now, I guess. You can replace it with 2 functions: firstLine(params) and secondLine(params)

Reply:
@vlad20012 done